source/SkiaSharp.HarfBuzz/SkiaSharp.HarfBuzz.Shared/SKShaper.cs

```diff
@@ -67,6 +68,14 @@ public Result Shape(Buffer buffer, float xOffset, float yOffset, SKPaint paint)
 			float textSizeY = paint.TextSize / FONT_SIZE_SCALE;
 			float textSizeX = textSizeY * paint.TextScaleX;
 
+			// adjust alignment
+			if (paint.TextAlign != SKTextAlign.Left) {
+				var width = pos.Sum(p => p.XAdvance) * textSizeX;
```

Comment:
This might be a huge performance hit. I think text alignment should be evaluated when the text blob is being drawn onto a canvas. Better implement this in the DrawShapedText extension method. SKTextBlob.Bounds should give you needed Size for your calculations.

Reply:
Hi, thanks for your suggestion. I tried to implement it, but Bounds return much larger box than there is actually rendered.

For example:
`SKTextBlob.Create("test", font).Width`
returns 157.3125
while 
`paint.MeasureText("test")`
returns 67.125

According to the documentation of bounds:
```
    /** Returns conservative bounding box. Uses SkPaint associated with each glyph to
        determine glyph bounds, and unions all bounds. Returned bounds may be
        larger than the bounds of all glyphs in runs.

        @return  conservative bounding box
    */
```

When diving deeper into the code of `SkTextBlobBuilder`, there are two functions to calculate the bounds, one is called `TightRunBounds` and the other `ConservativeRunBounds`, where `ConservativeRunBounds` is used for `sk_textblob_builder_alloc_run_pos` and `sk_textblob_builder_alloc_run_pos_h`, and `TightRunBounds` is only used for `sk_textblob_builder_alloc_run` (and there is a FIXME comment to not use tight run in this case too).
So it seems that bounds is pretty useless for this use case.

Another option is to use paint.MeasureText but it would not be accurate with things like Arabic text (which is the point of SKShaper)

Another option to that I can think of is to calculate the width during the loop of Shape (so there won't be any performance hit), and return it as part of the Result and than use it in DrawShapedText. What do you think?
Thanks.

Reply:
https://github.com/koolkdev/SkiaSharp/commit/17da8ad14534a9630fd3e192187d2edf9cbde971

Reply:
DrawPositionedText should use https://github.com/mono/SkiaSharp/blob/fffae82cf6ec1c4a6f490ddb780709ce9211725e/binding/Binding/SKTextBlob.cs#L806 if that doesn't work there is something wrong with the binding

Reply:
Hmm, I think `bounds` is misused here and should be an out parameter. Only when a nullptr is used Skia calculates the bounds.

Reply:
It seems that bounds is an input parameter, according to documentation:
```
        bounds defines an optional bounding box, used to suppress drawing when SkTextBlob
        bounds does not intersect SkSurface bounds. If bounds is nullptr, SkTextBlob bounds
        is computed from RunBuffer::pos, and RunBuffer::glyphs metrics.
```
Skia calculate the bounds anyway, but that parameter is just the initial value which it does SkRect.join to 
https://github.com/mono/skia/blob/916a1e9721de74a05c75391f52393972ed7f736c/src/core/SkTextBlob.cpp#L514
https://github.com/mono/skia/blob/916a1e9721de74a05c75391f52393972ed7f736c/src/core/SkTextBlob.cpp#L383
At least if I am understanding the code correctly....

Anyway, the issue is the that Skia uses conservative way to calculate the bounds (probably for performance) which result in a much larger bounds than the actual bounds. (like twice as big or more in some cases), so it doesn't really help in this case.

What do you say about the current solution?